src/dmd/statementsem.d

```diff
@@ -1670,306 +1765,241 @@ private extern (C++) final class StatementSemanticVisitor : Visitor
+            default:
+                assert(0);
+        }
+        const(char)* r = (fs.op == TOK.foreach_reverse_) ? "R" : "";
+        int j = sprintf(fdname.ptr, "_aApply%s%.*s%llu", r, 2, fntab[flag], cast(ulong)dim);
+        assert(j < BUFFER_LEN);
+
+        FuncDeclaration fdapply;
+        TypeDelegate dgty;
+        auto params = new Parameters();
+        params.push(new Parameter(STC.in_, tn.arrayOf(), null, null, null));
+        auto dgparams = new Parameters();
+        dgparams.push(new Parameter(0, Type.tvoidptr, null, null, null));
+        if (dim == 2)
+            dgparams.push(new Parameter(0, Type.tvoidptr, null, null, null));
+        dgty = new TypeDelegate(new TypeFunction(ParameterList(dgparams), Type.tint32, LINK.d));
+        params.push(new Parameter(0, dgty, null, null, null));
+        fdapply = FuncDeclaration.genCfunc(params, Type.tint32, fdname.ptr);
+
+        if (tabty == Tsarray)
+            fs.aggr = fs.aggr.castTo(sc2, tn.arrayOf());
+        // paint delegate argument to the type runtime expects
+        Expression fexp = flde;
+        if (!dgty.equals(flde.type))
+        {
+            fexp = new CastExp(loc, flde, flde.type);
+            fexp.type = dgty;
+        }
+        ec = new VarExp(Loc.initial, fdapply, false);
+        ec = new CallExp(loc, ec, fs.aggr, fexp);
+        ec.type = Type.tint32; // don't run semantic() on ec
+        return ec;
+    }
+    private static extern(D) Expression applyAssocArray(ForeachStatement fs, Expression flde, TypeAArray taa)
+    {
+        Expression ec;
+        const dim = fs.parameters.dim;
+        // Check types
+        Parameter p = (*fs.parameters)[0];
+        bool isRef = (p.storageClass & STC.ref_) != 0;
+        Type ta = p.type;
+        if (dim == 2)
+        {
+            Type ti = (isRef ? taa.index.addMod(MODFlags.const_) : taa.index);
+            if (isRef ? !ti.constConv(ta) : !ti.implicitConvTo(ta))
+            {
+                fs.error("`foreach`: index must be type `%s`, not `%s`",
+                         ti.toChars(), ta.toChars());
+                return null;
+            }
+            p = (*fs.parameters)[1];
+            isRef = (p.storageClass & STC.ref_) != 0;
+            ta = p.type;
+        }
+        Type taav = taa.nextOf();
+        if (isRef ? !taav.constConv(ta) : !taav.implicitConvTo(ta))
+        {
+            fs.error("`foreach`: value must be type `%s`, not `%s`",
+                     taav.toChars(), ta.toChars());
+            return null;
+        }
 
-        case Terror:
-            s = new ErrorStatement();
-            break;
+        /* Call:
+         *  extern(C) int _aaApply(void*, in size_t, int delegate(void*))
+         *      _aaApply(aggr, keysize, flde)
+         *
+         *  extern(C) int _aaApply2(void*, in size_t, int delegate(void*, void*))
+         *      _aaApply2(aggr, keysize, flde)
+         */
+        __gshared FuncDeclaration* fdapply = [null, null];
```

Comment:
why is this __gshared?

Reply:
No idea. was like that before. I think this is supposed to be a lazily computed constant. I'll do another PR to change e.g. https://github.com/dlang/dmd/pull/12298#discussion_r599362148 I'll have another look at this to see if there's a better way to do this.